### PR TITLE
fix(dist_reduce): run mesh all_reduce when DTensor mesh is orthogonal

### PR DESCRIPTION
**Pull request by saforem2**
Since #2963, `_dist_reduce` short-circuits all DTensor inputs with `return float(x.full_tensor().item())`, skipping the requested mesh all_reduce entirely. The justification in the comment ("avoid double-counting") only holds when the DTensor's own mesh covers the requested reduction mesh; when they are orthogonal — as in the trainer's loss reduction where the loss is a Replicated DTensor on the TP mesh and the requested mesh is `loss_mesh = batch x cp` — the cross-batch sum is silently dropped.

Symptom: every TP > 1 run reports `loss = true_loss / dp_world_size`. For a 2B run at TP=2 with 24 ranks, step-1 loss is 1.08 instead of the ~12.95 ln(vocab) baseline observed at TP=1. Gradients and optimizer steps are unaffected because backward goes through `loss.backward()` not `_dist_reduce` — only the printed/logged loss is wrong.

Fix: detect mesh axis overlap via `mesh_dim_names`. If the DTensor's mesh shares any axis with the requested mesh, `full_tensor()` already covers that axis and we skip the explicit all_reduce as before. If the meshes are disjoint, we materialize via `full_tensor()` and fall through to the plain-tensor path so the requested-mesh all_reduce runs.

Adds a unit test covering both the orthogonal case (was broken) and the same-mesh case (must not double-count).

### CODE REVIEW

File: torchtitan/distributed/utils.py [hunk 1]
```diff
@@ -45,23 +45,37 @@ def _dist_reduce(
...
+        # If the requested reduction mesh is the same mesh the DTensor
+        # already lives on, ``full_tensor()`` has done the reduction and
+        # we must skip the subsequent all_reduce to avoid double-counting.
+        # Otherwise (the requested mesh is orthogonal to the DTensor's mesh,
+        # e.g. reducing a TP-Replicated loss across batch_mesh), we still
+        # need to reduce on the requested mesh — fall through to the
+        # plain-tensor path with the materialized full tensor.
+        x_mesh = x.device_mesh
+        x = x.full_tensor()
+        if mesh is None or _meshes_share_axes(x_mesh, mesh):
```
Comment (tianyu-l) [inline torchtitan/distributed/utils.py:68]: 1. do we need to handle the `mesh is None` specially here?
2. when `_meshes_share_axes` is True, it doesn't mean the result is correct? but you would just return. I think a workaround is to just error out when two meshes are not identical.

Comment (fegin) [inline torchtitan/distributed/utils.py:68]: I want to argue that the only clean workaround is to_local(). Then we don't need to infer whether it is full dtensor or non-full dtensor + TP. The actual fix is to deprecate the non-full dtensor case. Then we will unconditionally apply full_tensor() here. But given that it is hard to do deprecation without full dtensor being ready, we should just make it a plain tensor, which was what we always did for the TP case before this April -- we changed the TP-enabled model to always use DTensor for activation and loss this April.

I fixed this issue in https://github.com/pytorch/torchtitan/pull/3159. If you want to land with this PR, I prefer landing `to_local()`.

cc., @tianyu-l

**Review by fegin — CHANGES_REQUESTED** [review]
I prefer using `to_local()` for now to fix the issue: https://github.com/pytorch/torchtitan/pull/3159.

**Comment by saforem2** [comment]
Closing as superseded.

The orthogonal-mesh under-reporting was fixed by @fegin in #3159
(commit `d64eabcce`, merged 2026-05-18) — switching the DTensor
branch of `_dist_reduce` from `full_tensor()` to `to_local()` + the
explicit mesh `all_reduce`. From that PR's issue list:

> Our `_dist_reduce` under-reports loss for non full tensor:
> `full_tensor()` reduces only over the DTensor's own mesh (TP-only
> in legacy), skipping DP. Loss was under-reported by
> `dp_shard × cp`. Switched to `to_local()` then mesh all_reduce
> over loss_mesh.

That's the same case the #3204 tests covered (Replicate-on-TP loss
reduced across an orthogonal batch mesh). Verified on current
`upstream/main`: `_dist_reduce` now does `x = x.to_local()` for the
DTensor path, then runs the requested `funcol.all_reduce` —
no double-count, no skipped reduction.

@fegin / @tianyu-l — thanks for the review and pointer to #3159.
The `to_local()` shape is cleaner than the `full_tensor()` +
mesh-overlap helper I had here, agreed.
